src/ripple/rpc/impl/GRPCHelpers.cpp

```diff
@@ -0,0 +1,1854 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/rpc/impl/GRPCHelpers.h>
+
+namespace ripple {
+namespace RPC {
+
+// In the below populateProto* functions, getProto is a function that returns
+// a reference to the mutable protobuf message to be populated. The reason this
+// is a function, as opposed to just a pointer or reference to the object,
+// is that there is no way to get a non-const reference, and getting a pointer
+// to the proto object causes default initialization of the object. However,
+// if the corresponding field is not present in the STObject, we don't want to
+// initialize the proto object. To get around this, getProto is a function that
+// is called only if the field is present in the STObject
+template <class T>
+void
+populateProtoU8(STObject const& obj, SF_U8 const& field, T const& getProto)
```

Comment:
I think we can replace these eight fuctions:
```
populateProtoU8
populateProtoU16
populateProtoU32
populateProtoU64
populateProtoH128
populateProtoH160
populateProtoH256
populateProtoVL
```
with this:
```
template <class T, class L>
void
populateProtoPrimitive(STObject const& from, TypedField<T> const& field, L&& getProto)
{
    if (!from.isFieldPresent(field))
        return;

    if constexpr (std::is_integral_v<typename T::value_type>)
    {
        getProto()->set_value(from[field]);
    }
    else
    {
        auto v = from[field];
        getProto()->set_value(v.data(), v.size());
    }
}
```
(please dummy check me). I don't object if you want to keep the original versions tho.

Reply:
I think this will work. I'll continue to think about it, but as of now I actually prefer your method